File: node/tests/AddressResolver.test.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 */

import { afterAll, beforeAll, describe, expect, it } from "@jest/globals";
import { ValkeyCluster } from "../../utils/TestUtils.js";
import { GlideClient, GlideClusterClient, ProtocolVersion } from "../build-ts";
import {
    getClientConfigurationOption,
    getServerVersion,
    parseEndpoints,
} from "./TestUtilities";

const TIMEOUT = 50000;

describe("AddressResolver", () => {
    let standaloneCluster: ValkeyCluster;
    let clusterCluster: ValkeyCluster;

    beforeAll(async () => {
        const standaloneAddresses: string =
            global.STAND_ALONE_ENDPOINT as string;
        standaloneCluster = standaloneAddresses
            ? await ValkeyCluster.initFromExistingCluster(
                  false,
                  parseEndpoints(standaloneAddresses),
                  getServerVersion,
              )
            : await ValkeyCluster.createCluster(false, 1, 1, getServerVersion);

        const clusterAddresses: string = global.CLUSTER_ENDPOINTS as string;
        clusterCluster = clusterAddresses
            ? await ValkeyCluster.initFromExistingCluster(
                  true,
                  parseEndpoints(clusterAddresses),
                  getServerVersion,
              )
            : await ValkeyCluster.createCluster(true, 3, 1, getServerVersion);
    }, 120000);

    afterAll(async () => {
        if (standaloneCluster) {
            await standaloneCluster.close();
        }

        if (clusterCluster) {
            await clusterCluster.close();
        }
    });

    it(
        "address resolver with fake address - standalone",
        async () => {
            const addresses = standaloneCluster.getAddresses();
            const [actualHost, actualPort] = addresses[0];

            const fakeHost = "fake-host-that-does-not-exist.invalid";
            const fakePort = 9999;

            // Resolver translates fake address to real server address
            const resolver = (host: string, port: number): [string, number] => {
                if (host === fakeHost && port === fakePort) {
                    return [actualHost, actualPort];
                }

                return [host, port];
            };

            const client = await GlideClient.createClient({
                ...getClientConfigurationOption(
                    [[fakeHost, fakePort]],
                    ProtocolVersion.RESP3,
                ),
                addressResolver: resolver,
            });

            try {
                const pingResult = await client.ping();
                expect(pingResult).toBe("PONG");

                const setResult = await client.set(
                    "resolver_test_key",
                    "resolver_test_value",
                );
                expect(setResult).toBe("OK");

                const getResult = await client.get("resolver_test_key");
                expect(getResult).toBe("resolver_test_value");
            } finally {
                client.close();
            }
        },
        TIMEOUT,
    );

    it(
        "address resolver with fake address - cluster",
        async () => {
            const addresses = clusterCluster.getAddresses();
            const [actualHost, actualPort] = addresses[0];

            const fakeHost = "fake-cluster-host.invalid";
            const fakePort = 9999;

            // Resolver translates fake address to real server address
            const resolver = (host: string, port: number): [string, number] => {
                if (host === fakeHost && port === fakePort) {
                    return [actualHost, actualPort];
                }

                return [host, port];
            };

            const client = await GlideClusterClient.createClient({
                ...getClientConfigurationOption(
                    [[fakeHost, fakePort]],
                    ProtocolVersion.RESP3,
                ),
                addressResolver: resolver,
            });

            try {
                const pingResult = await client.ping();
                expect(pingResult).toBe("PONG");

                const setResult = await client.set(
                    "cluster_resolver_test_key",
                    "cluster_resolver_test_value",
                );
                expect(setResult).toBe("OK");

                const getResult = await client.get("cluster_resolver_test_key");
                expect(getResult).toBe("cluster_resolver_test_value");
            } finally {
                client.close();
            }
        },
        TIMEOUT,
    );

    it(
        "address resolver exception falls back to original - standalone",
        async () => {
            const addresses = standaloneCluster.getAddresses();
            const [actualHost, actualPort] = addresses[0];

            // Resolver that always throws - params are consumed to satisfy lint
            const resolver = (host: string, port: number): [string, number] => {
                void host;
                void port;
                throw new Error("test-exception");
            };

            // Client should still connect using the original (valid) address as fallback
            const client = await GlideClient.createClient({
                ...getClientConfigurationOption(
                    [[actualHost, actualPort]],
                    ProtocolVersion.RESP3,
                ),
                addressResolver: resolver,
            });

            try {
                const pingResult = await client.ping();
                expect(pingResult).toBe("PONG");
            } finally {
                client.close();
            }
        },
        TIMEOUT,
    );

    it(
        "address resolver exception falls back to original - cluster",
        async () => {
            const addresses = clusterCluster.getAddresses();
            const [actualHost, actualPort] = addresses[0];

            // Resolver that always throws - params are consumed to satisfy lint
            const clusterResolver = (
                host: string,
                port: number,
            ): [string, number] => {
                void host;
                void port;
                throw new Error("test-exception");
            };

            // Client should still connect using the original (valid) address as fallback
            const client = await GlideClusterClient.createClient({
                ...getClientConfigurationOption(
                    [[actualHost, actualPort]],
                    ProtocolVersion.RESP3,
                ),
                addressResolver: clusterResolver,
            });

            try {
                const pingResult = await client.ping();
                expect(pingResult).toBe("PONG");
            } finally {
                client.close();
            }
        },
        TIMEOUT,
    );
});

File: go/integTest/address_resolver_test.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package integTest

import (
	"context"

	glide "github.com/valkey-io/valkey-glide/go/v2"
	"github.com/valkey-io/valkey-glide/go/v2/config"
)

func (suite *GlideTestSuite) TestAddressResolverWithFakeAddress_Standalone() {
	// Get the actual server address from test configuration
	actualHost := suite.standaloneHosts[0].Host
	actualPort := suite.standaloneHosts[0].Port

	// Use a fake/placeholder address in configuration
	fakeHost := "fake-host-that-does-not-exist.invalid"
	fakePort := 9999

	// Create resolver that translates fake address to real server address
	resolver := func(host string, port int) (string, int) {
		if host == fakeHost && port == fakePort {
			return actualHost, actualPort
		}
		return host, port
	}

	// Configure client with fake address - connection should succeed because
	// resolver translates it to the real address
	clientConfig := defaultClientConfig().
		WithAddress(&config.NodeAddress{Host: fakeHost, Port: fakePort}).
		WithAddressResolver(resolver)

	client, err := glide.NewClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	// Verify the client works - this proves the resolved address was used
	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)

	setResult, err := client.Set(ctx, "resolver_test_key", "resolver_test_value")
	suite.Require().NoError(err)
	suite.Equal("OK", setResult)

	getResult, err := client.Get(ctx, "resolver_test_key")
	suite.Require().NoError(err)
	suite.Equal("resolver_test_value", getResult.Value())
}

func (suite *GlideTestSuite) TestAddressResolverWithFakeAddress_Cluster() {
	// Get the actual server address from test configuration
	actualHost := suite.clusterHosts[0].Host
	actualPort := suite.clusterHosts[0].Port

	// Use a fake/placeholder address in configuration
	fakeHost := "fake-cluster-host.invalid"
	fakePort := 9999

	// Create resolver that translates fake address to real server address
	resolver := func(host string, port int) (string, int) {
		if host == fakeHost && port == fakePort {
			return actualHost, actualPort
		}
		return host, port
	}

	// Configure client with fake address - connection should succeed because
	// resolver translates it to the real address
	clientConfig := defaultClusterClientConfig().
		WithAddress(&config.NodeAddress{Host: fakeHost, Port: fakePort}).
		WithAddressResolver(resolver)

	client, err := glide.NewClusterClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	// Verify the client works - this proves the resolved address was used
	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)

	setResult, err := client.Set(ctx, "cluster_resolver_test_key", "cluster_resolver_test_value")
	suite.Require().NoError(err)
	suite.Equal("OK", setResult)

	getResult, err := client.Get(ctx, "cluster_resolver_test_key")
	suite.Require().NoError(err)
	suite.Equal("cluster_resolver_test_value", getResult.Value())
}

func (suite *GlideTestSuite) TestAddressResolverExceptionFallsBackToOriginal_Standalone() {
	// Get the actual server address from test configuration
	actualHost := suite.standaloneHosts[0].Host
	actualPort := suite.standaloneHosts[0].Port

	// Create resolver that always panics
	resolver := func(host string, port int) (string, int) {
		panic("test-exception")
	}

	// Configure client with real address - connection should still succeed
	// because the fallback uses the original address when resolver panics
	clientConfig := defaultClientConfig().
		WithAddress(&config.NodeAddress{Host: actualHost, Port: actualPort}).
		WithAddressResolver(resolver)

	client, err := glide.NewClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)
}

func (suite *GlideTestSuite) TestAddressResolverExceptionFallsBackToOriginal_Cluster() {
	// Get the actual server address from test configuration
	actualHost := suite.clusterHosts[0].Host
	actualPort := suite.clusterHosts[0].Port

	// Create resolver that always panics
	resolver := func(host string, port int) (string, int) {
		panic("test-exception")
	}

	// Configure client with real address - connection should still succeed
	// because the fallback uses the original address when resolver panics
	clientConfig := defaultClusterClientConfig().
		WithAddress(&config.NodeAddress{Host: actualHost, Port: actualPort}).
		WithAddressResolver(resolver)

	client, err := glide.NewClusterClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)
}

func (suite *GlideTestSuite) TestAddressResolverReturnsEmptyFallsBackToOriginal_Standalone() {
	// Get the actual server address from test configuration
	actualHost := suite.standaloneHosts[0].Host
	actualPort := suite.standaloneHosts[0].Port

	// Create resolver that returns empty host (signals fallback)
	resolver := func(host string, port int) (string, int) {
		return "", 0
	}

	// Configure client with real address - connection should still succeed
	// because returning empty/zero signals fallback to original address
	clientConfig := defaultClientConfig().
		WithAddress(&config.NodeAddress{Host: actualHost, Port: actualPort}).
		WithAddressResolver(resolver)

	client, err := glide.NewClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)
}

func (suite *GlideTestSuite) TestAddressResolverReturnsEmptyFallsBackToOriginal_Cluster() {
	// Get the actual server address from test configuration
	actualHost := suite.clusterHosts[0].Host
	actualPort := suite.clusterHosts[0].Port

	// Create resolver that returns empty host (signals fallback)
	resolver := func(host string, port int) (string, int) {
		return "", 0
	}

	// Configure client with real address - connection should still succeed
	// because returning empty/zero signals fallback to original address
	clientConfig := defaultClusterClientConfig().
		WithAddress(&config.NodeAddress{Host: actualHost, Port: actualPort}).
		WithAddressResolver(resolver)

	client, err := glide.NewClusterClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)
}

func (suite *GlideTestSuite) TestAddressResolverNil_Standalone() {
	// Get the actual server address from test configuration
	actualHost := suite.standaloneHosts[0].Host
	actualPort := suite.standaloneHosts[0].Port

	// Configure client with nil resolver - should work normally
	clientConfig := defaultClientConfig().
		WithAddress(&config.NodeAddress{Host: actualHost, Port: actualPort}).
		WithAddressResolver(nil)

	client, err := glide.NewClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)
}

func (suite *GlideTestSuite) TestAddressResolverNil_Cluster() {
	// Get the actual server address from test configuration
	actualHost := suite.clusterHosts[0].Host
	actualPort := suite.clusterHosts[0].Port

	// Configure client with nil resolver - should work normally
	clientConfig := defaultClusterClientConfig().
		WithAddress(&config.NodeAddress{Host: actualHost, Port: actualPort}).
		WithAddressResolver(nil)

	client, err := glide.NewClusterClient(clientConfig)
	suite.Require().NoError(err)
	defer client.Close()

	ctx := context.Background()
	result, err := client.Ping(ctx)
	suite.Require().NoError(err)
	suite.Equal("PONG", result)
}
